### PR TITLE
react: Fix compile error

### PR DESCRIPTION
**Pull request by ghost**
This fixes the test `const htmlElement = React.createElement("input", { type: "text" });`.
There is probably a better way to fix this?

### CODE REVIEW

**Comment by tkrotoff** [comment]
I don't understand why [`extends React.Component<LinkProps>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/048b0b6aa1af4f7505e78a3ffb311d002489ec01/types/react-app/index.d.ts#L38) from react-app fails in Travis CI with error `Generic type 'Component<P, S>' requires 2 type argument(s)`

**Comment by vsiao** [comment]
This can be fixed by adding a string-literal override to `createElement` which references `InputHTMLAttributes<T>`.

**Comment by andy-ms** [comment]
@vsiao Thanks! Do you know if `& ClassAttributes<T>` is still necessary or should `InputHTMLAttributes<T>` alone be good?

File: types/react/index.d.ts [hunk 1]
```diff
@@ -186,7 +186,7 @@ declare namespace React {
     // DOM Elements
     function createElement<P extends HTMLAttributes<T>, T extends HTMLElement>(
         type: keyof ReactHTML,
-        props?: ClassAttributes<T> & P,
+        props?: InputHTMLAttributes<T> & ClassAttributes<T> & P,
```
Comment (vsiao) [inline types/react/index.d.ts]: this isn't quite right—I meant something like:

```ts
function createElement(
    type: "input",
    props?: InputHTMLAttributes<HTMLInputElement> & ClassAttributes<HTMLInputElement>,
    ...
```

as an _additional_ overload for `createElement`.

If we're thinking more generally, ideally there would be an interface that maps each DOM type to its attributes interface, like

```ts
interface DOMAttributesMap {
    input: InputHTMLAttributes<HTMLInputElement> & ClassAttributes<HTMLInputElement>,
    ...
}
```

and then we can type `createElement` with lookup types:

```ts
function createElement<T extends keyof DOMAttributesMap>(
    type: T,
    props?: DOMAttributesMap[T],
    ...
```

but we don't have to tackle that in this PR.

File: types/react/index.d.ts [hunk 2]
```diff
@@ -184,6 +184,11 @@ declare namespace React {
     function createFactory<P>(type: ComponentClass<P>): Factory<P>;
 
     // DOM Elements
+    // TODO: generalize this to everything in `keyof ReactHTML`, not just "input"
+    function createElement(
+        type: "input",
+        props?: InputHTMLAttributes<HTMLInputElement> & ClassAttributes<HTMLInputElement>,
+        ...children: ReactNode[]): DetailedReactHTMLElement<{}, HTMLInputElement>;
```
Comment (vsiao) [inline types/react/index.d.ts]: Should be `DetailedReactHTMLElement<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>`, I think.

**Review by vsiao — APPROVED** [review]
one small thing, otherwise looks good!

**Comment by johnnyreilly** [comment]
Thanks!

**Pull request merged**